main.py

```diff
@@ -96,6 +96,15 @@ def main_game(screen, width, height, player, nodes, paths):
     for node in nodes:
         node.draw(screen)
 
+    for node in nodes:
```

Comment:
this is great perhaps you should wait for @mahdishahsavar to merge his code to see the main game logic that we are trying to adhere too. BUt great job on creating a class to abstract your problem/challenge.

Now we can easily import it in the main game logic at will. But these changes you made to main.py I believe will break a lot of things for other challenges added previously. 

Sync with @mahdishahsavar  about this. Thanks

Reply:
This is good Hannah, I'll try to write it in the same format as other challenges check it whenever you have time Hannah so we can use it in the main besides other challenges.

Reply:
`import pygame
import socket

class IPAddressChallenge:
    def __init__(self):
        pygame.init()
        self.id = "IP_challenge"
        self.screen_width = 1000
        self.screen_height = 800
        self.screen = pygame.display.set_mode((self.screen_width, self.screen_height))
        pygame.display.set_caption('IP Address Challenge')
        self.font = pygame.font.Font(None, 20)
        self.initialize_interface_elements()

    def initialize_interface_elements(self):
        self.WHITE = (255, 255, 255)
        self.BLACK = (0, 0, 0)
        self.GRAY = (200, 200, 200)
        self.instruction_text = [
            "Instructions:",
            "1. You've received a message from an anonymous source about a vulnerable server.",
            "2. The message reads: 'Check the ports on 10.0.0.1'. Write a script to scan ports.",
            "3. Click RUN when you are done to test your script."
        ]
        self.user_code = "def scan_ports(target_ip):\n    # Write your port scanning logic here\n"
        self.run_button = self.render_run_button()

    def render_run_button(self):
        button = pygame.Rect(30, self.screen_height - 60, 75, 30)
        pygame.draw.rect(self.screen, self.BLACK, button)
        self.render_text_to_surface("RUN", [button.x + 5, button.y + 5], self.font, self.WHITE)
        return button

    def render_text_to_surface(self, text, position, font, color):
        text_surface = font.render(text, True, color)
        self.screen.blit(text_surface, position)

    def validate_user_code(self, user_code):
        local_vars = {}
        try:
            exec(user_code, {"__builtins__": {}}, local_vars)
            result = local_vars.get('scan_ports', lambda x: None)("10.0.0.1")
            # Assume the function returns a list of open ports
            return 'Open ports: ' + ', '.join(map(str, result))
        except Exception as e:
            return f"Error executing your script: {str(e)}"

    def run(self):
        running = True
        while running:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    running = False
                elif event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_RETURN:
                        self.user_code += '\n'
                    else:
                        self.user_code += event.unicode
                elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
                    if self.run_button.collidepoint(pygame.mouse.get_pos()):
                        print(self.validate_user_code(self.user_code))

            self.screen.fill(self.WHITE)
            self.render_text_to_surface('\n'.join(self.instruction_text), (20, 20), self.font, self.BLACK)
            pygame.draw.rect(self.screen, self.GRAY, self.run_button)
            self.render_run_button()
            pygame.display.flip()

if __name__ == '__main__':
    challenge = IPAddressChallenge()
    challenge.run()
`